Implemented a purge policy for Old DWH snapshots

chandrashekar-s opened the pull request:
The following changes have been made 
 - Implemented a periodic purge scheduler
 - The purge job will scan for the DWH snapshots and retain only the last N successfully completed snapshots. The previously completed DWH snapshots will be deleted. Snapshots which are incomplete will be ignored by this job i.e. they wont be considered for deletion.
 - For successful state of a DWH snapshot we now create two timestamp files (1) timestamp_start.txt at the beginning of the pipeline run (2) timestamp_end.txt at the end of the pipeline run, as opposed to only one file (timestamp.txt) which we used to create earlier.
 - On start of the application we now initialise to the most recent successfully completed snapshot (which checks for the presence of both the timestamp files).

## Description of what I changed

<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->
<!--- If your PR is related to an issue , please mention it here. -->
<!--- It is generally a good practice to first file an issue with enough
  context and reference it in the PR, but if you don't have that, please remove
  this section. -->

## E2E test

The changes have been tested end to end locally for the below scenarios.
 1. Created few DWH snapshots by running the batch pipeline and incremental runs, checked for the trigger of the purge job and verified if only the last N runs are retained as per the configuration.
 2. On application restart checked if it is initialised to the most recent successfully completed snapshot which is the last-1 snapshot since the latest one was incomplete (had manually deleted the timestamp_end.txt file for the recent DWH).
 3. Also verified if the job ignores incomplete snapshots.

TESTED:

Tested the full runs and incremental runs for the Unix file systems as well as GCS file systems. The number of snapshots retained are manually verified after the purge job completions. 

## Checklist: I completed these to help reviewers :)

<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->

- [x] I have read and will follow the [review process](https://github.com/GoogleCloudPlatform/openmrs-fhir-analytics/blob/master/doc/review_process.md).
- [x] I am familiar with Google Style Guides for the language I have coded in.

  No? Please take some time and review [Java](https://google.github.io/styleguide/javaguide.html) and [Python](https://google.github.io/styleguide/pyguide.html) style guides.

- [x] My IDE is configured to follow the Google [**code styles**](https://google.github.io/styleguide/).

  No? Unsure? -> [configure your IDE](https://github.com/google/google-java-format).

- [x] I have **added tests** to cover my changes. (If you refactored existing code that was well tested you do not have to add tests)
- [x] I ran `mvn clean package` right before creating this pull request and added all formatting changes to my commit.
- [x] All new and existing **tests passed**.
- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+
+  private LocalDateTime lastPurgeRunEnd;
+
+  private volatile boolean isPurgeJobRunning;
```
Does this save us a lot? Why not avoid `volatile` and use `synchronized` like everywhere else (I mean even in this class you still need to use `synchronized`)?

chandrashekar-s replied:
We could make the `purgeDwhFiles` method synchronized, which means any two threads trying to access this method for the same object will get serialised (one will wait for the other to get completed). Since this is a periodic job and the throughput is less, synchronisation should be fine. However, just to make to little optimal I introduced this lock behaviour so that whichever thread gets the lock will execute the job and the other thread will simply end (no waiting).

bashir2 replied:
I see, it is fine if you don't want to change `purgeDwhFiles` to `synchronized` but access to `isPurgeJobRunning` is `synchronized` already, right? I mean, just dropping `volatile` here does not hurt any concurrent logic, does it?

chandrashekar-s replied:
You are right, dropping the volatile keyword does not impact anything. Have done the changes.

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+  }
+
+  private synchronized boolean lockPurgeJob() {
+    if (isPurgeJobRunning()) {
```
nit: Suggestion, remove set/get methods for `isPurgeJobRunning` (given they are `private`) and directly use the field; I don't think those methods help much with readability (and make the code longer).

chandrashekar-s replied:
Done

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+    String rootPrefix = dataProperties.getDwhRootPrefix();
+    Preconditions.checkState(rootPrefix != null && !rootPrefix.isEmpty());
+
+    int lastNSnapshotsToRetain = dataProperties.getLastNDWHSnapshotsToRetain();
```
nit: `lastNSnapshotsToRetain` -> `numSnapshotsToRetain`

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+
+    int lastNSnapshotsToRetain = dataProperties.getLastNDWHSnapshotsToRetain();
+    Preconditions.checkState(
+        lastNSnapshotsToRetain > 0, "lastNSnapshotsToRetain should be positive value");
```
Let's move this check to `init()` such that it is triggered in `@PostConstruct`.

chandrashekar-s replied:
Done

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+   * ones. Any incomplete snapshots will not be purged by this job and will have to be manually
+   * verified and acted upon.
+   */
+  private void purgeDWHFiles() {
```
Question: Have you looked into how these methods can be unit-tested, e.g., using an in memory file-system? Our file handling logic in the controller is becoming more complex and I am wondering how hard it is to start writing unit-tests for these.

chandrashekar-s replied:
I have added a few unit tests for the LocalFileSystem, this is written by manually handling the creation and deletion of  directories + files. Will explore on the in-memory file system and get back.

bashir2 replied:
Okay, thanks for adding the new unit-tests. Please feel free to resolve this conversation if adding tests for the GCS case is too much work.

chandrashekar-s replied:
Have added few test cases for GCS case as well, resolving this for now.

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+      deleteOlderSnapshots(paths, recentSnapshotsToBeRetained);
+      setLastPurgeRunStatus(PurgeRunStatus.SUCCESS);
+    } catch (IOException e) {
+      setLastPurgeRunStatus(PurgeRunStatus.FAILURE);
```
Does this call have any affect?

chandrashekar-s replied:
Yes, this has no effect. I have refactored the code now to not have these lines.

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+
+  private void setLastPurgeRunStatus(PurgeRunStatus status) {
+    if (status == PurgeRunStatus.SUCCESS) {
+      lastPurgeRunEnd = LocalDateTime.now();
```
IIUC, this means that if one run fails (for whatever reason) we will attempt to run again almost immediately (i.e., within 30 seconds), correct? I would guess that it is likely that the next run also fails and given that this is an IO bound task, this can become a performance issue. Should we increase the 30 seconds check interval to a larger value?

chandrashekar-s replied:
Yes, you are right that if the purge job fails then immediately in the next run (which is 30 sec from now) the job will be retried. Shall I increase it to 10 mins?

chandrashekar-s replied:
Update the periodic timer to 5 minutes.

bashir2 replied:
Sounds good, thanks.

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+          // and their parent directories can be successfully deleted.
+          allDirectoryPathSet.add(
+              currentDirectory
+                  .resolve(".temp-beam", StandardResolveOptions.RESOLVE_DIRECTORY)
```
What happens if `.temp-beam` does not exist in `currentDirectory`?

chandrashekar-s replied:
If there is no `.temp-beam` directory in the `currentDirectory` then the `FileSystems.delete(..)` api ignores them (internally the beam is catching and ignoring the exception, check [here](https://github.com/apache/beam/blob/26f5792c8cc7f0e2a04a3b94c57a96feabf92156/sdks/java/core/src/main/java/org/apache/beam/sdk/io/LocalFileSystem.java#L206)).

bashir2 replied:
Great, thanks for checking.

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+      }
+    }
+    List<String> allDirectoryPathList = new ArrayList<>(allDirectoryPathSet);
+    allDirectoryPathList.sort(Collections.reverseOrder());
```
Why do we need the sort here?

chandrashekar-s replied:
Sorting is done so that, first the leaf child directories gets deleted, followed by their corresponding parent directories. Added a comment in the code explaining the same.

bashir2 replied:
SG; thanks.

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+   * @return the status of DWH completeness.
+   * @throws IOException
+   */
+  public boolean isDWHComplete(ResourceId dwhResource) throws IOException {
```
nit: In general in names use `Dwh` instead of `DWH` (and for any other abbreviation).

chandrashekar-s replied:
Done

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+   * @param dwhRootPrefix
+   * @return the base directory name
+   */
+  public String getBaseDir(String dwhRootPrefix) {
```
I think for this and some of the other methods in this class, we don't really need `public` and package-private should be enough; if yes, please drop `public` where it is not needed.

chandrashekar-s replied:
This method is used by the PipelineManager.java as well, hence this is public. Also, made sure only the methods which are externally used are made public.

bashir2 replied:
Right but those are all in the same *package* i.e., `org.openmrs.analytics`, right? So I think package-private should still work. Not a big deal but I usually use "public" when I expect usage from outside my package.

chandrashekar-s replied:
Agreed, have made the methods as package-private.

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+   * @return The list of all child directories under the base directory
+   * @throws IOException
+   */
+  public Set<ResourceId> getAllChildDirectories(String baseDir) throws IOException {
```
For the few functions that are moved from `PipelineManager`, I assumed no code is changed; please let me know if that is not the case.

chandrashekar-s replied:
Yes, the methods which are moved from `PipelineManager' are not modified.

chandrashekar-s replied:
I have made small refactoring to the methods `getBaseDir` and `getPrefix` which are moved from `PipelineManager', to made them little robust for different file systems.

bashir2 replied:
Okay, thanks.

bashir2 reviewed (commented):
Thanks @chandrashekar-s for this change.

chandrashekar-s commented on pipelines/controller/config/application.yaml:
```diff
@@ -36,6 +36,15 @@ fhirdata:
   # Note a too frequent run might be too resource intensive.
   incrementalSchedule: "0 0 * * * *"
 
+  # The schedule format is similar to Spring's CronExpression, i.e.,
```
Done

chandrashekar-s commented on pipelines/controller/config/application.yaml:
```diff
@@ -36,6 +36,15 @@ fhirdata:
+  # Note a too frequent run might be too resource intensive.
+  purgeSchedule: "0 0 * * * *"
+
+  lastNDWHSnapshotsToRetain: 5
```
Done

chandrashekar-s commented on pipelines/controller/config/application.yaml:
```diff
@@ -36,6 +36,15 @@ fhirdata:
+  # "0 0 * * * *" means top of every hour;
+  # "*/40 * * * * *" means every 40 seconds;
+  # Note a too frequent run might be too resource intensive.
+  purgeSchedule: "0 0 * * * *"
```
Done

chandrashekar-s commented on pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java:
```diff
@@ -0,0 +1,353 @@
+/*
```
Done

chandrashekar-s commented:
@bashir2 Thanks for the review. I have addressed the review comments and checked in the changes. Have added few unit tests in the new commit, more to be added especially for GCS file system scenario. Will explore on the option of in-memory file system for test cases.

chandrashekar-s commented:
In the latest commit, I have improved the test coverage and refactored couple of existing methods to make them little robust for different file systems.

bashir2 commented on pipelines/controller/src/main/java/org/openmrs/analytics/DwhFilesManager.java:
```diff
@@ -335,4 +356,39 @@ public Set<ResourceId> getAllChildDirectories(String baseDir) throws IOException
+      case GcsPath.SCHEME:
+        String gcsObject = GcsPath.fromUri(dwhRootPrefix).getObject();
+        int position = gcsObject.lastIndexOf("/");
+        if (position != -1) index = dwhRootPrefix.indexOf(gcsObject) + position;
```
Just checking: what happens if the `dwhRootPrefix` is `gs://my_bucket/my_object_prefix_`? In this case, does `GcsPath.fromUri(dwhRootPrefix).getObject()` returns `/my_object_prefix` or just `my_object_prefix`? If the latter, then doesn't that lead to a bug (index being `-1`)?

chandrashekar-s replied:
If `dwhRootPrefix` is `gs://my_bucket/my_object_prefix_` then `GcsPath.fromUri(dwhRootPrefix).getObject()` return `my_object_prefix`. This leads to index being -1 and throws an error. We are handling this error as we are expecting the gcs url to be of the format `gs://<bucket>/<baseDirPath>/<prefix>`, if `baseDirPath` is empty then application throws an error.

bashir2 reviewed (approved):
Thanks @chandrashekar-s most of the comments are questions or small suggestions, please feel free to merge this after addressing them. Also please make sure that the purge functionality works with GCS files and update the PR description `TESTED` field with that.

chandrashekar-s commented on pipelines/controller/src/main/java/org/openmrs/analytics/DwhFilesManager.java:
```diff
@@ -240,15 +231,15 @@ private TreeSet<String> getRecentSnapshots(List<ResourceId> paths, int numberOfS
 
   /**
-   * This method tells us whether the DWH snapshot has been successfully completed by the pipeline.
+   * This method tells us whether the dwh snapshot has been successfully completed by the pipeline.
```
Ok. Have made the changes accordingly.

chandrashekar-s commented on pipelines/controller/src/main/java/org/openmrs/analytics/DwhFilesManager.java:
```diff
@@ -335,4 +356,39 @@ public Set<ResourceId> getAllChildDirectories(String baseDir) throws IOException
+    return index;
+  }
+
+  private static String parseScheme(String spec) {
```
Ok. Have used [URI.getScheme](https://docs.oracle.com/javase/8/docs/api/java/net/URI.html#getScheme--) now.

chandrashekar-s commented:
Thanks @bashir2 for the review, there is one unresolved comment for which I have provided my comments. Can you please check and let me know if that's fine.
